{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Pull protein sequences from NCBI\n",
    "Prior to this, DAVID (DAVID Functional Annotation Bioinformatics Microarray Analysis, https://david.ncifcrf.gov/conversion.jsp ) was used to convert REF_SEQ accession numbers to Protein_GI_Accession numbers. **GI_number_generator_AK.ipynb**\n",
    "\n",
    "After this, percent match between the input protein sequence/Entrez ID will be compared to each of the generated PROTEIN_seq."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>GENENAME</th>\n",
       "      <th>ORG</th>\n",
       "      <th>ENTREZID</th>\n",
       "      <th>REFSEQ</th>\n",
       "      <th>LOG2FC</th>\n",
       "      <th>True_EID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>ABC_transporter</td>\n",
       "      <td>PA</td>\n",
       "      <td>879411</td>\n",
       "      <td>NP_248876</td>\n",
       "      <td>-0.038469</td>\n",
       "      <td>15595384</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ABC_transporter</td>\n",
       "      <td>PA</td>\n",
       "      <td>883108</td>\n",
       "      <td>NP_248894</td>\n",
       "      <td>-0.207718</td>\n",
       "      <td>15595400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>ABC_transporter</td>\n",
       "      <td>PA</td>\n",
       "      <td>878380</td>\n",
       "      <td>NP_249014</td>\n",
       "      <td>-0.187309</td>\n",
       "      <td>15595520</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>ABC_transporter</td>\n",
       "      <td>PA</td>\n",
       "      <td>880771</td>\n",
       "      <td>NP_249293</td>\n",
       "      <td>0.085173</td>\n",
       "      <td>15595799</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>ABC_transporter</td>\n",
       "      <td>PA</td>\n",
       "      <td>879023</td>\n",
       "      <td>NP_249295</td>\n",
       "      <td>0.038834</td>\n",
       "      <td>15595801</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0         GENENAME ORG  ENTREZID     REFSEQ    LOG2FC  True_EID\n",
       "0           0  ABC_transporter  PA    879411  NP_248876 -0.038469  15595384\n",
       "1           1  ABC_transporter  PA    883108  NP_248894 -0.207718  15595400\n",
       "2           2  ABC_transporter  PA    878380  NP_249014 -0.187309  15595520\n",
       "3           3  ABC_transporter  PA    880771  NP_249293  0.085173  15595799\n",
       "4           4  ABC_transporter  PA    879023  NP_249295  0.038834  15595801"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# import libraries and define associated email for queries\n",
    "from Bio import Entrez\n",
    "import time\n",
    "import pandas as pd\n",
    "\n",
    "Entrez.email =\"karanjia@uw.edu\" # maybe switch this to a user input\n",
    "\n",
    "data = pd.read_csv('GI_Numbers_combo.csv')\n",
    "data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "numpy.int64"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(data['True_EID'][0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">GI  728886557  AIY94952 flagellar hook-basal body protein FlhP [Bacillus subtilis subsp. subtilis str. 168]\n",
      "mlrsmltasttlnqlqqqidtissnlsnsnttgykakdtnfselvrqqfdqvdekneevakarktppglrlgvgammssrlvsdqgsiqktdrdldiaftspyqylqvnvngnrqytrdgalyvtpsaananqlqlvtgngypvldengntvnidssmknitinkngtltasdgnavqrfnlgvvqvnnpqelksegnnlfsidnaaafeelnganrqnigmqqgslemsnvdiseqmtdlitsqrsyqlnsrtitmgdqmlglinsvr\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GI</th>\n",
       "      <th>PROTEIN_SEQ</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [GI, PROTEIN_SEQ]\n",
       "Index: []"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# define gis and use Biopython.Entrez to pull protein sequences from NCBI\n",
    "\n",
    "gis = [728886557] # gi = gi number, input is PROTEIN_GI\n",
    "                         # eventually replace with column from correct df \n",
    "request = Entrez.epost(\"protein\",id=\",\".join(map(str,gis)))\n",
    "result = Entrez.read(request)\n",
    "webEnv = result[\"WebEnv\"]\n",
    "queryKey = result[\"QueryKey\"]\n",
    "handle = Entrez.efetch(db=\"protein\",retmode=\"xml\", webenv=webEnv, query_key=queryKey)\n",
    "\n",
    "d = [] # empty dataframe \n",
    "\n",
    "for r in Entrez.parse(handle):\n",
    "    try:\n",
    "        gi=int([x for x in r['GBSeq_other-seqids'] if \"gi\" in x][0].split(\"|\")[1])\n",
    "    except ValueError:\n",
    "        gi=None\n",
    "    print (\">GI \",gi,\" \" + r[\"GBSeq_primary-accession\"] + \" \" + r[\"GBSeq_definition\"] + \"\\n\" + r[\"GBSeq_sequence\"])\n",
    "    #print (\">GI \",gi,\" \" +  r[\"GBSeq_sequence\"])\n",
    "    #d.append((gi, r['GBSeq_sequence']))\n",
    "    \n",
    "pd.DataFrame(d, columns=('GI', 'PROTEIN_SEQ'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "id_list = data[\"True_EID\"]\n",
    "\n",
    "d = []\n",
    "\n",
    "#for i in range(len(id_list)):\n",
    "request = Entrez.epost(\"protein\",id=\",\".join(map(str,id_list)))\n",
    "result = Entrez.read(request)\n",
    "webEnv = result[\"WebEnv\"]\n",
    "queryKey = result[\"QueryKey\"]\n",
    "handle = Entrez.efetch(db=\"protein\",retmode=\"xml\", webenv=webEnv, query_key=queryKey)\n",
    "\n",
    "for r in Entrez.parse(handle):\n",
    "    try:\n",
    "        gi=int([x for x in r['GBSeq_other-seqids'] if \"gi\" in x][0].split(\"|\")[1])\n",
    "    except ValueError:\n",
    "        gi=None\n",
    "    d.append((r['GBSeq_sequence']))\n",
    "    \n",
    "prot_seq = pd.DataFrame(d, columns=['PROTEIN_SEQ'])\n",
    "prot_seq.to_csv('prot_seq.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PROTEIN_SEQ</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>mkaltssllglfaapvlagllgayvplasaappkeiriavpdvsag...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>mhqriasiglgltlalggsaqaagqlnvvswsgyfspqllekfeke...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mtyrtpltllfaaglalggqaraegtlhfanwsdyyppellkkfek...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>mlpamrtgllcallgvtapawaeyvtvisfggankeaqetafykpf...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>mskslkaaslkfatlaaglacaaqamavdltvvsfgganksaqika...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>431</th>\n",
       "      <td>mtqpskttkltkdevdrlisdyqtkqdeqaqetlvrvytnlvdmla...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>432</th>\n",
       "      <td>mdmklqqvqvlkpqltqelrqaitllgyhsaelaeyidelslenpl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>433</th>\n",
       "      <td>mdvevkkngknaqlkdhevkelikqsqngdqqardllieknmrlvw...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>434</th>\n",
       "      <td>msrnkveicgvdtsklpvlkneemrklfrqlqdegddsareklvng...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>435</th>\n",
       "      <td>mkklklrlthlwykllmklglksdevyyiggsealppplskdeeqv...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>436 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           PROTEIN_SEQ\n",
       "0    mkaltssllglfaapvlagllgayvplasaappkeiriavpdvsag...\n",
       "1    mhqriasiglgltlalggsaqaagqlnvvswsgyfspqllekfeke...\n",
       "2    mtyrtpltllfaaglalggqaraegtlhfanwsdyyppellkkfek...\n",
       "3    mlpamrtgllcallgvtapawaeyvtvisfggankeaqetafykpf...\n",
       "4    mskslkaaslkfatlaaglacaaqamavdltvvsfgganksaqika...\n",
       "..                                                 ...\n",
       "431  mtqpskttkltkdevdrlisdyqtkqdeqaqetlvrvytnlvdmla...\n",
       "432  mdmklqqvqvlkpqltqelrqaitllgyhsaelaeyidelslenpl...\n",
       "433  mdvevkkngknaqlkdhevkelikqsqngdqqardllieknmrlvw...\n",
       "434  msrnkveicgvdtsklpvlkneemrklfrqlqdegddsareklvng...\n",
       "435  mkklklrlthlwykllmklglksdevyyiggsealppplskdeeqv...\n",
       "\n",
       "[436 rows x 1 columns]"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prot_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>GENENAME</th>\n",
       "      <th>ORG</th>\n",
       "      <th>ENTREZID</th>\n",
       "      <th>REFSEQ</th>\n",
       "      <th>LOG2FC</th>\n",
       "      <th>True_EID</th>\n",
       "      <th>SEQUENCE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>ABC_transporter</td>\n",
       "      <td>PA</td>\n",
       "      <td>879411</td>\n",
       "      <td>NP_248876</td>\n",
       "      <td>-0.038469</td>\n",
       "      <td>15595384</td>\n",
       "      <td>mkaltssllglfaapvlagllgayvplasaappkeiriavpdvsag...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ABC_transporter</td>\n",
       "      <td>PA</td>\n",
       "      <td>883108</td>\n",
       "      <td>NP_248894</td>\n",
       "      <td>-0.207718</td>\n",
       "      <td>15595400</td>\n",
       "      <td>mhqriasiglgltlalggsaqaagqlnvvswsgyfspqllekfeke...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>ABC_transporter</td>\n",
       "      <td>PA</td>\n",
       "      <td>878380</td>\n",
       "      <td>NP_249014</td>\n",
       "      <td>-0.187309</td>\n",
       "      <td>15595520</td>\n",
       "      <td>mtyrtpltllfaaglalggqaraegtlhfanwsdyyppellkkfek...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>ABC_transporter</td>\n",
       "      <td>PA</td>\n",
       "      <td>880771</td>\n",
       "      <td>NP_249293</td>\n",
       "      <td>0.085173</td>\n",
       "      <td>15595799</td>\n",
       "      <td>mlpamrtgllcallgvtapawaeyvtvisfggankeaqetafykpf...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>ABC_transporter</td>\n",
       "      <td>PA</td>\n",
       "      <td>879023</td>\n",
       "      <td>NP_249295</td>\n",
       "      <td>0.038834</td>\n",
       "      <td>15595801</td>\n",
       "      <td>mskslkaaslkfatlaaglacaaqamavdltvvsfgganksaqika...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>431</th>\n",
       "      <td>431</td>\n",
       "      <td>RNA_polymerase</td>\n",
       "      <td>BS</td>\n",
       "      <td>939937</td>\n",
       "      <td>NP_388354</td>\n",
       "      <td>-0.055848</td>\n",
       "      <td>728883360</td>\n",
       "      <td>mtqpskttkltkdevdrlisdyqtkqdeqaqetlvrvytnlvdmla...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>432</th>\n",
       "      <td>432</td>\n",
       "      <td>RNA_polymerase</td>\n",
       "      <td>BS</td>\n",
       "      <td>936362</td>\n",
       "      <td>NP_391300</td>\n",
       "      <td>-0.140629</td>\n",
       "      <td>728886343</td>\n",
       "      <td>mdmklqqvqvlkpqltqelrqaitllgyhsaelaeyidelslenpl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>433</th>\n",
       "      <td>433</td>\n",
       "      <td>RNA_polymerase</td>\n",
       "      <td>BS</td>\n",
       "      <td>938729</td>\n",
       "      <td>NP_390226</td>\n",
       "      <td>-0.528350</td>\n",
       "      <td>728885268</td>\n",
       "      <td>mdvevkkngknaqlkdhevkelikqsqngdqqardllieknmrlvw...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>434</th>\n",
       "      <td>434</td>\n",
       "      <td>RNA_polymerase</td>\n",
       "      <td>BS</td>\n",
       "      <td>939953</td>\n",
       "      <td>NP_389416</td>\n",
       "      <td>0.303125</td>\n",
       "      <td>728884442</td>\n",
       "      <td>msrnkveicgvdtsklpvlkneemrklfrqlqdegddsareklvng...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>435</th>\n",
       "      <td>435</td>\n",
       "      <td>RNA_polymerase</td>\n",
       "      <td>BS</td>\n",
       "      <td>939683</td>\n",
       "      <td>NP_389415</td>\n",
       "      <td>-0.083972</td>\n",
       "      <td>728884441</td>\n",
       "      <td>mkklklrlthlwykllmklglksdevyyiggsealppplskdeeqv...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>436 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Unnamed: 0         GENENAME ORG  ENTREZID     REFSEQ    LOG2FC  \\\n",
       "0             0  ABC_transporter  PA    879411  NP_248876 -0.038469   \n",
       "1             1  ABC_transporter  PA    883108  NP_248894 -0.207718   \n",
       "2             2  ABC_transporter  PA    878380  NP_249014 -0.187309   \n",
       "3             3  ABC_transporter  PA    880771  NP_249293  0.085173   \n",
       "4             4  ABC_transporter  PA    879023  NP_249295  0.038834   \n",
       "..          ...              ...  ..       ...        ...       ...   \n",
       "431         431   RNA_polymerase  BS    939937  NP_388354 -0.055848   \n",
       "432         432   RNA_polymerase  BS    936362  NP_391300 -0.140629   \n",
       "433         433   RNA_polymerase  BS    938729  NP_390226 -0.528350   \n",
       "434         434   RNA_polymerase  BS    939953  NP_389416  0.303125   \n",
       "435         435   RNA_polymerase  BS    939683  NP_389415 -0.083972   \n",
       "\n",
       "      True_EID                                           SEQUENCE  \n",
       "0     15595384  mkaltssllglfaapvlagllgayvplasaappkeiriavpdvsag...  \n",
       "1     15595400  mhqriasiglgltlalggsaqaagqlnvvswsgyfspqllekfeke...  \n",
       "2     15595520  mtyrtpltllfaaglalggqaraegtlhfanwsdyyppellkkfek...  \n",
       "3     15595799  mlpamrtgllcallgvtapawaeyvtvisfggankeaqetafykpf...  \n",
       "4     15595801  mskslkaaslkfatlaaglacaaqamavdltvvsfgganksaqika...  \n",
       "..         ...                                                ...  \n",
       "431  728883360  mtqpskttkltkdevdrlisdyqtkqdeqaqetlvrvytnlvdmla...  \n",
       "432  728886343  mdmklqqvqvlkpqltqelrqaitllgyhsaelaeyidelslenpl...  \n",
       "433  728885268  mdvevkkngknaqlkdhevkelikqsqngdqqardllieknmrlvw...  \n",
       "434  728884442  msrnkveicgvdtsklpvlkneemrklfrqlqdegddsareklvng...  \n",
       "435  728884441  mkklklrlthlwykllmklglksdevyyiggsealppplskdeeqv...  \n",
       "\n",
       "[436 rows x 8 columns]"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data['SEQUENCE'] = pd.DataFrame(d)\n",
    "\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "data.to_csv('compiled_prot_seq.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Cells below this markdown are rough drafts of the previous set of cells, keep for documentation of different options: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "Entrez.email = \"karanjia@uw.edu\"\n",
    "\n",
    "\n",
    "def retrieve_annotation(id_list):\n",
    "\n",
    "    \"\"\"Annotates Entrez Gene IDs using Bio.Entrez, in particular epost (to\n",
    "    submit the data to NCBI) and esummary to retrieve the information.\n",
    "    Returns a list of dictionaries with the annotations.\"\"\"\n",
    "\n",
    "    request = Entrez.epost(\"gene\", id=\",\".join(id_list))\n",
    "    try:\n",
    "        result = Entrez.read(request)\n",
    "    except RuntimeError as e:\n",
    "        # FIXME: How generate NAs instead of causing an error with invalid IDs?\n",
    "        print (\"An error occurred while retrieving the annotations.\")\n",
    "        print (\"The error returned was %s\") % e\n",
    "        sys.exit(-1)\n",
    "\n",
    "    webEnv = result[\"WebEnv\"]\n",
    "    queryKey = result[\"QueryKey\"]\n",
    "    data = Entrez.esummary(db=\"gene\", webenv=webEnv, query_key=queryKey)\n",
    "    annotations = Entrez.read(data)\n",
    "\n",
    "    print (\"Retrieved %d annotations for %d genes\") % (len(annotations), len(id_list))\n",
    "\n",
    "    return annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "sequence item 0: expected str instance, int found",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-23-558ba55f6c9b>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mretrieve_annotation\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m879411\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m883108\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m<ipython-input-21-36ffa567c095>\u001b[0m in \u001b[0;36mretrieve_annotation\u001b[0;34m(id_list)\u001b[0m\n\u001b[1;32m      8\u001b[0m     Returns a list of dictionaries with the annotations.\"\"\"\n\u001b[1;32m      9\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 10\u001b[0;31m     \u001b[0mrequest\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mEntrez\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mepost\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"gene\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mid\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\",\"\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mid_list\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     11\u001b[0m     \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     12\u001b[0m         \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mEntrez\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrequest\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mTypeError\u001b[0m: sequence item 0: expected str instance, int found"
     ]
    }
   ],
   "source": [
    "retrieve_annotation([879411, 883108])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['187']\n"
     ]
    }
   ],
   "source": [
    "# don't use this\n",
    "from Bio import Entrez\n",
    "\n",
    "def entry_ENTREZ(entrezID): \n",
    "    handle = Entrez.efetch(db='nucleotide', id=entrezID, rettype = 'fasta', retmode= 'text')\n",
    "    f = open('%s.fasta' % singleID, 'w')\n",
    "    f.write(handle.read())\n",
    "    handle.close()\n",
    "    f.close()\n",
    "\n",
    "#get an id list: this makes a big search and gets a list of id \n",
    "handle = Entrez.esearch(db='genome', term=\"Pseudomonas aeruginosa[Orgn]\", idtype=\"acc\")\n",
    "record = Entrez.read(handle)\n",
    "handle.close()\n",
    "print(record[\"IdList\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
